### PR TITLE
allow pytest --migrations to succeed

### PR DESCRIPTION
**Pull request by chrismeyersfsu**
##### SUMMARY
We actually subvert migrations from running in test via pytest.ini --no-migrations option. This has led to bit rot for the sqlite migrations happy path. This changeset pays off that tech debt and allows for an sqlite migration happy path.
* This paves the way for programatic invocation of individual migrations and weaving of the creation of resources (i.e. Instance, Job Template, etc). With this, a developer can instantiate various database states, trigger a migration, assert the state of the db, and then have pytest rollback all of that.
* I will note that in practice, running these migrations is dog shit slow BUT this work also opens up the possibility of saving and re-using sqlite3 database files. Normally, caching is not THE answer and causes more harm than good. But in this case, our migrations are mostly write-once (I say mostly because this change set violates that :) so cache invalidation isn't a major issue.ge does.

* Adds new ci test `api-migration`

##### ISSUE TYPE
 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME
 - API

##### AWX VERSION

```

```

##### ADDITIONAL INFORMATION

To test this. Run the below without these changes. See everything blow up?. Now run the tests with these changes ... success.
```
pytest --migrations
```

Note: you can also just run like 1 test. You just want to trigger the running of migrations to bring up the test db. Oh, and you _might_ need to delete the test db file `rm -f awx/awx_test.sqlite3`

### CODE REVIEW

File: awx/main/migrations/_sqlite_helper.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            raise ValueError("sqlite_code parameter required")
+        sqlite_code = kwargs.pop('sqlite_code')
+
+        self.sqlite_code = sqlite_code
+        self.sqlite_reverse_code = kwargs.pop('sqlite_reverse_code', None)
+        super().__init__(*args, **kwargs)
+
+    def database_forwards(self, app_label, schema_editor, from_state, to_state):
+        if not schema_editor.connection.vendor.startswith('postgres'):
+            self.code = self.sqlite_code or migrations.RunPython.noop
```
Comment (AlanCoding) [inline awx/main/migrations/_sqlite_helper.py:47]: I don't love this behavior. One point, months ago, I did the same thing you did here - made the migrations compatible with sqlite3. But for the life of me I can not find the branch where I worked on that.

I just added a condition in the migration logic depending on the vendor, and that didn't take as many changes as you could think. Sure there are lots of logic steps we just don't want to run, but doing this I expect we're going to miss some stuff we should have, like creating default system jobs, or creating default credential types.

Also the naming of `sqlitemigrations` this is kind of concerning when taken out of this context.

Comment (chrismeyersfsu) [reply awx/main/migrations/_sqlite_helper.py:47]: I don't follow.

Do you not like that the user can supply `None` for `sqlite_code` and are advocating for requiring `sqlite_code` to always be provided?
OR
Do you not like the overall idea of postgres sql vs. sqlite sql?
OR
Do you not like the variable choice that we are pivoting on?
OR
Do you not like "burying" this pivot logic and would rather the pivot be directly in the developers face, in the migration?

I'm fine with renaming `sqlitemigrations`

Comment (chrismeyersfsu) [reply awx/main/migrations/_sqlite_helper.py:47]: Talked with @AlanCoding offline.

TL;DR `self.code = self.sqlite_code or self.code or migrations.RunPython.noop` instead of `self.code = self.sqlite_code or migrations.RunPython.noop` :+1:

File: awx/main/migrations/0113_v370_event_bigint.py [hunk 2]
```diff
@@ -24,6 +26,11 @@ def migrate_event_data(apps, schema_editor):
             cursor.execute(f'ALTER TABLE {tblname} ALTER COLUMN id TYPE bigint USING id::bigint;')
 
 
+def migrate_event_data_sqlite(apps, schema_editor):
+    # TODO: cmeyers fill this in
```
Comment (AlanCoding) [inline awx/main/migrations/0113_v370_event_bigint.py:30]: Still got a TODO here. Why? Are you really going to migrate event data? I see no way you will get the hourly event tables to work in sqlite3. Of course it doesn't support range partitioning from the get-go, but even if you created superficial "partitions" as separate tables, I'm skeptical that it could be made to work, or that it would be worth it if you could do it.

Comment (fosterseth) [inline awx/main/migrations/0113_v370_event_bigint.py:30]: is something needed here?

Comment (AlanCoding) [reply awx/main/migrations/0113_v370_event_bigint.py:30]: Oh, this is for the BigInt migration. I'm _sure_ we won't ever want to do this in sqlite3. The next file does migration to partitions, which again, may be feasible superficially, but questionable value.

File: awx/main/migrations/0185_move_JSONBlob_to_JSONField.py [hunk 3]
```diff
@@ -104,11 +106,12 @@ class Migration(migrations.Migration):
             name='deleted_actor',
             field=models.JSONField(null=True),
         ),
-        migrations.RunSQL(
+        dbawaremigrations.RunSQL(
             """
             ALTER TABLE main_activitystream RENAME setting TO setting_old;
             ALTER TABLE main_activitystream ALTER COLUMN setting_old DROP NOT NULL;
             """,
+            sqlite_sql="ALTER TABLE main_activitystream RENAME setting TO setting_old",
```
Comment (AlanCoding) [inline awx/main/migrations/0185_move_JSONBlob_to_JSONField.py:114]: ok I guess this is kind of nice.

**Comment by chrismeyersfsu** [comment]
I might should remove the todo. I’m not going to do it in this or.

A noop seems fine for the cases I have a todo for the time being. If someone makes a change that requires testing code and data around those todo then they may have to pay the penalty of making them work in SQLite.

File: awx/main/migrations/_sqlite_helper.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    def database_forwards(self, app_label, schema_editor, from_state, to_state):
+        if not schema_editor.connection.vendor.startswith('postgres'):
+            self.sql = self.sqlite_sql or migrations.RunSQL.noop
+        super().database_forwards(app_label, schema_editor, from_state, to_state)
+
+    def database_backwards(self, app_label, schema_editor, from_state, to_state):
+        if not schema_editor.connection.vendor.startswith('postgres'):
+            self.reverse_sql = self.sqlite_reverse_sql or migrations.RunSQL.noop
+        super().database_forwards(app_label, schema_editor, from_state, to_state)
```
Comment (fosterseth) [inline awx/main/migrations/_sqlite_helper.py]: is this intended?
```suggestion
        super().database_backwards(app_label, schema_editor, from_state, to_state)
```

File: awx/main/migrations/_sqlite_helper.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+    def database_forwards(self, app_label, schema_editor, from_state, to_state):
+        if not schema_editor.connection.vendor.startswith('postgres'):
+            self.code = self.sqlite_code or self.code or migrations.RunPython.noop
+        super().database_forwards(app_label, schema_editor, from_state, to_state)
+
+    def database_backwards(self, app_label, schema_editor, from_state, to_state):
+        if not schema_editor.connection.vendor.startswith('postgres'):
+            self.reverse_code = self.sqlite_reverse_code or migrations.RunPython.noop
+        super().database_forwards(app_label, schema_editor, from_state, to_state)
```
Comment (fosterseth) [inline awx/main/migrations/_sqlite_helper.py]:
```suggestion
        super().database_backwards(app_label, schema_editor, from_state, to_state)
```

File: awx/main/migrations/_sqlite_helper.py (hunk 4 above)
Comment (chrismeyersfsu) [reply awx/main/migrations/_sqlite_helper.py]: Nope, not intentional. Thank you!

**Review by AlanCoding — APPROVED** [review]
